README.md
[![Travis Build](https://travis-ci.org/Ks89/My-MEAN-website-client.svg?branch=master)](https://travis-ci.org/Ks89/My-MEAN-website-client)   [![AppVeyor Build](https://ci.appveyor.com/api/projects/status/x7r2v139hi84cvsj/branch/master?svg=true)](https://ci.appveyor.com/project/Ks89/my-mean-website-server/branch/master)   [![Code Climate](https://codeclimate.com/github/Ks89/My-MEAN-website-client/badges/gpa.svg)](https://codeclimate.com/github/Ks89/My-MEAN-website-client)   [![CodeClimate Coverage](https://codeclimate.com/github/Ks89/My-MEAN-website-client/badges/coverage.svg)](https://codeclimate.com/github/Ks89/My-MEAN-website-client/coverage)   [![Coveralls Coverage](https://coveralls.io/repos/github/Ks89/My-MEAN-website-client/badge.svg?branch=master)](https://coveralls.io/github/Ks89/My-MEAN-website-client?branch=master)   [![Known Vulnerabilities](https://snyk.io/test/github/ks89/my-mean-website-client/badge.svg)](https://snyk.io/test/github/ks89/my-mean-website-client)

<br>
# My M.E.A.N. website client/front-end (Alpha)
<br>
**This is the client side.** Server side is available [HERE](https://github.com/Ks89/My-MEAN-website-server)
<br>
## Informations
My MEAN website is a MEAN's web application that I'm creating as a personal website, but also for other users.
It's composed by:
- A: a front-end in Angular 2
- N + E: a back-end in Node.js + Express js (and other useful libs like PassportJs)
- M: a MongoDb's database
- redis
- webpack + gulp
- and other stuff

A possible extension of this project will be a configurable template to build a custom web app very quickly. This is my final goal, please be patient :)

Attention! This project is still an alpha, so it's not production ready. Please be careful.
If you are interested, star this project on GitHub.

Testing:
- front-end unit: coverage >80%
- front-end e2e: work in progress, In future alphas I'll complete everything

## Requirements
- Node.js and npm
- My-MEAN-website server Alpha4 or greater
- npm
- PhantomJS, Google Chrome and Firefox (mandatory for testing)
- some global npm dependencies, for instance karma-cli, mocha, webpack, typescript, gulp 4.0 alpha, remap-istanbul, webdriver-manager, protractor
- work in progress... (this is only an alpha, please be patient)


## News
- *12/30/2016* - **My MEAN website** Alpha 4 public release [HERE](https://github.com/Ks89/My-MEAN-website-client/releases/tag/v.alpha-4)
- *11/28/2016* - **My MEAN website** Alpha 3 public release [HERE](https://github.com/Ks89/My-MEAN-website-client/releases/tag/v.alpha-3)
- *10/27/2016* - **My MEAN website** Alpha 2 public release [HERE](https://github.com/Ks89/My-MEAN-website-server/releases/tag/v.alpha-2.2)
- *08/15/2016* - **My MEAN website** Alpha 1 public release [HERE](https://github.com/Ks89/My-MEAN-website-server/releases/tag/v.alpha-1)

## How to install (MacOS)
- from the `setup` folder of this project, run `bash install-macos.sh`

## How to install (Linux)
- from the `setup` folder of this project, run `bash install-linux.sh`

## How to install (Windows)
- install Node.js, MongoDb, redis-server and so on
- from the `setup` folder of this project, run `bash install-windows.sh`
- TODO improve this tutorial :)

## How to setup
1. install all necessary tools
2. execute `npm install` into the root folder
3. start My-MEAN-website server with `gulp`
7. execute `npm start` into the root folder to start this application (client/front-end)

This will start this application at http://localhost:3300

## How to run tests (client-side)
Execute `npm test`

## How to start
- cd MyWebsite (if necessary rename server-side's main folder)
- `gulp`
- cd ..
- cd 'main folder of this project'
- `npm start` (or `npm run build` for production)
- Open your browser **http://localhost:3300** (if not automatically opened)

If you want to start the admin page, go to **http://localhost:3300/admin.html**
Attention: it's very unstable, because I decided to implement only some basic features. I'll improve it in the next alphas.

## Features
Work in progress... (this is only an alpha, please be patient)


## Future extensions
Work in progress... (this is only an alpha, please be patient)


## Images

![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/home.png)
<br/><br/>
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/projects.png)
<br/><br/>
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/projectDetail.png)
<br/><br/>
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/projectDetail-image.png)
<br/><br/>
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/contact.png)
<br/><br/>
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/contact-images.png)
<br/><br/>
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/signin.png)
<br/><br/>
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/register.png)
<br/><br/>
Note: updated local profile info
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/profile-updated.png)
<br/><br/>
Note: multiple account connected (Local, Facebook, Github, Google and Linkedin)
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/profile-multiple.png)
<br/><br/>
Admin web page (list of users)
![alt tag](http://www.stefanocappa.it/publicfiles/Github_repositories_images/MyMeanWebsite/admin-users.png)


## Configuration
Work in progress... (this is only an alpha, please be patient)


## Thanks
A special thanks to the authors of these books, because very useful to understand how to develop a modern web application: [BOOK1](https://www.manning.com/books/getting-mean-with-mongo-express-angular-and-node) and [BOOK2](https://www.manning.com/books/angular-2-development-with-typescript). Also to [this project](https://github.com/AngularClass/angular2-webpack-starter)


## License

Copyright 2015-2017 Stefano Cappa

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

<br/>
**Created by Stefano Cappa**


NEWS.md
For news, upcoming features and so on, refers to https://github.com/Ks89/My-MEAN-website-client/milestones

app.js
console.log("Starting with NODE_ENV=" + process.env.NODE_ENV);
console.log("process.env.CI is " + process.env.CI);

if(!(process.env.CI && process.env.CI === 'yes')) {
  console.log("Initializing dotenv (requires .env file)")
  require('dotenv').config(); //to read info from .env file
  //attention: i'm using "dotenv" 2.0 and for this reason I must call "config()".
}

var path = require('path');

// --------------------------------------------------------
// --------------------------------------------------------
// See this issue here https://github.com/Ks89/My-MEAN-website/issues/30
//  to understand this piece of code.
var pathFrontEndFolder, pathFrontEndIndex;
var pathFrontEndAdminIndex;
if((process.env.CI && process.env.CI === 'yes') || process.env.NODE_ENV === 'test') {
  console.log("Executed in CI or TEST - providing fake app_client and index.html");
  //provides fake directories and files to be able to run this files
  //also with mocha in both testing and ci environments.
  //Otherwise, you are forced to run `npm run build` into app_client's folder
  pathFrontEndFolder = path.join(__dirname);
  pathFrontEndIndex = path.join(__dirname, 'app.js');
} else {
  console.log("Providing reale app_client and index.html");
  pathFrontEndFolder = path.join(__dirname, 'app_client', 'dist');
  pathFrontEndIndex = path.join(__dirname, 'app_client', 'dist', 'index.html');
  pathFrontEndAdminIndex = path.join(__dirname, 'app_client', 'dist', 'admin.html');
}
// --------------------------------------------------------
// --------------------------------------------------------

var express = require('express');
var compression = require('compression');
var favicon = require('serve-favicon');
var morgan = require('morgan');
var session = require('express-session');
var bodyParser = require('body-parser');
//logger created with winston
var logger = require("./app_server/utils/logger");

var redis   = require("redis"); //it's really useful?
var RedisStore = require('connect-redis')(session);
var client  = redis.createClient(); //it's really useful?

// --------------------------------------------------------------------------
// ----------------------------security packages-----------------------------
// --------------------------------------------------------------------------
// All security features are prefixed with `--SEC--`
// --SEC-- - github analog-nico/hpp [NOT helmet]
//    [http params pollution] security package to prevent http params pollution
var hpp = require('hpp');
// --SEC-- - [CSRF] github.com/expressjs/csurf [NOT helmet]
var csrf = require('csurf');
// --SEC-- - authentication local/3dparty (OAuth)
var passport = require('passport');
// --SEC-- - github ericmdantas/express-content-length-validator [NOT helmet]
//    large payload attacks - Make sure this application is
//    not vulnerable to large payload attacks
var contentLength = require('express-content-length-validator');
const MAX_CONTENT_LENGTH_ACCEPTED = 9999; // constants used with `contentLength`
// --SEC-- - Helmet
var helmet = require('helmet');
// --------------------------------------------------------------------------
// --------------------------------------------------------------------------

console.log("Initializing mongodb");
//require for mongo
require('./app_server/models/db');
require('./app_server/controllers/authentication/passport')(passport);

console.log("Initializing expressjs");
var app = express();

console.log("Initializing helmet");
// --SEC-- - [helmet] enable helmet
// this automatically add 9 of 11 security features
/*
  -dnsPrefetchControl controls browser DNS prefetching
  -frameguard to prevent clickjacking
  -hidePoweredBy to remove the X-Powered-By header
  -hpkp for HTTP Public Key Pinning
  -hsts for HTTP Strict Transport Security
  -ieNoOpen sets X-Download-Options for IE8+
  -noSniff to keep clients from sniffing the MIME type
  -xssFilter adds some small XSS protections
*/
// The other features NOT included by default are:
/*
  -contentSecurityPolicy for setting Content Security Policy
  -noCache to disable client-side caching => I don't want this for better performances
  -referrerPolicy to hide the Referer header
*/
app.use(helmet());

// --SEC-- - noCache to disable client-side caching [helmet]
// I don't want this for better performances (leave commented :))
// app.use(helmet.noCache())

// --SEC-- - referrer-policy to hide the Referer header [helmet]
app.use(helmet.referrerPolicy({ policy: 'no-referrer' }))

// --SEC-- - Public Key Pinning (hpkp): HTTPS certificates can be forged,
//    allowing man-in-the middle attacks.
//    HTTP Public Key Pinning aims to help that. [helmet]
const ninetyDaysInSeconds = 7776000;
app.use(helmet.hpkp({
  maxAge: ninetyDaysInSeconds,
  sha256s: ['AbCdEf123=', 'ZyXwVu456='],
  includeSubdomains: true,         // optional
  reportUri: 'https://example.com/hpkp-report', // optional
  reportOnly: false,               // optional
  // Set the header based on a condition.
  setIf: (req, res)  => req.secure //optional ()
}));

// --SEC-- - Content Security Policy (CSP): Trying to prevent Injecting anything
//    unintended into our page. That could cause XSS vulnerabilities,
//    unintended tracking, malicious frames, and more. [helmet]
app.use(helmet.contentSecurityPolicy({
  directives: {
    defaultSrc: ["'self'", 'localhost:3000', 'localhost:3001', 'www.google.com', 'www.youtube.com'],
    scriptSrc: ["'self'", "'unsafe-inline'", "'unsafe-eval'", 'maxcdn.bootstrapcdn.com',
                'ajax.googleapis.com', 'cdnjs.cloudflare.com',
                'code.jquery.com', 'www.google.com',
                'www.gstatic.com'],
    styleSrc: ["'self'", 'ajax.googleapis.com', 'maxcdn.bootstrapcdn.com', 'cdnjs.cloudflare.com', "'unsafe-inline'"],
    fontSrc: ["'self'", 'maxcdn.bootstrapcdn.com'],
    imgSrc: ["'self'", 'localhost:3000', 'localhost:3001',
              'placehold.it', 'placeholdit.imgix.net', 'camo.githubusercontent.com',
              's3.amazonaws.com', 'cdnjs.cloudflare.com'],
    sandbox: ['allow-forms', 'allow-scripts', 'allow-same-origin', 'allow-popups'],
    frameSrc : ["'self'", 'www.google.com', 'www.youtube.com'], //frame-src is deprecated
    childSrc : ["'self'", 'www.google.com', 'www.youtube.com'],
    connectSrc: [
        "'self'", "cdnjs.cloudflare.com", "ajax.googleapis.com",
        "ws://localhost:3000", "ws://localhost:3001", "ws://localhost:3100",
        "ws://localhost:3300"
    ],
    reportUri: '/report-violation',
    objectSrc: ["'none'"]
  },
  // Set to true if you only want browsers to report errors, not block them
  reportOnly: false,
  // Set to true if you want to blindly set all headers: Content-Security-Policy,
  // X-WebKit-CSP, and X-Content-Security-Policy.
  setAllHeaders: false,
  // Set to true if you want to disable CSP on Android where it can be buggy.
  disableAndroid: false,
  // Set to false if you want to completely disable any user-agent sniffing.
  // This may make the headers less compatible but it will be much faster.
  // This defaults to 'true'.
  browserSniff: true
}));

// --SEC-- - large payload attacks:
//   this line enables the middleware for all routes [NOT helmet]
app.use(contentLength.validateMax({max: MAX_CONTENT_LENGTH_ACCEPTED,
  status: 400, message: "stop it!"})); // max size accepted for the content-length


console.log("Initializing morgan (logger)");

// uncomment after placing your favicon in /public
//app.use(favicon(path.join(__dirname, 'public', 'favicon.ico')));
app.use(morgan({ "stream": logger.stream }));

console.log("Initializing static resources");
app.use(express.static(pathFrontEndFolder));

console.log("Initializing bodyparser");

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

console.log("Initializing hpp");
// --SEC-- - http params pollution: activate http parameters pollution
// use this ALWAYS AFTER app.use(bodyParser.urlencoded()) [NOT helmet]
app.use(hpp());

console.log("Initializing Express session");
// Express Session
app.use(session({
    secret: 'keyboard cat',
    resave: false,
    saveUninitialized: true,
    store: new RedisStore({ host: 'localhost', port: 6379, client: client,ttl :  260}),
    // cookie: {
    //   httpOnly: false,
    //     secure: false, //to use true, you must use https. If you'll use true with http it won't work.
    //     //maxAge: 2160000000
    // }
}));

console.log("Initializing passportjs");

app.use(passport.initialize());
app.use(passport.session());

// compress all requests using gzip
app.use(compression());

console.log("Initializing REST apis and CSRF");

// --------------------------------------- ROUTES ---------------------------------------
// dedicated routes for angular logging with stacktracejs
// these router aren't protected with csrf, because declared before app.use(csrf()).
var loggerApi = require('./app_server/routes/log-api')(express);
app.use('/api/log', loggerApi);

// enable middleware CSRF by csurf package [NOT helmet]
// before app.use('/api', routesApi); to protect their,
// but after session and/or cookie initialization
app.use(csrf());
app.use(function (req, res, next) {
  res.cookie('XSRF-TOKEN', req.csrfToken());
  res.locals.csrftoken = req.csrfToken();
  next();
});

// APIs for all route protected with CSRF (all routes except for angular log's service)
var routesApi = require('./app_server/routes/index')(express);
app.use('/api', routesApi);
// --------------------------------------------------------------------------------------

console.log("Initializing static path for index.html");

app.use('/', function(req, res) {
  res.sendFile(pathFrontEndIndex);
});

app.use('/admin', function(req, res) {
  res.sendFile(pathFrontEndAdminIndex);
});

// catch bad csrf token
app.use(function (err, req, res, next) {
  if (err.code !== 'EBADCSRFTOKEN') {
    return next(err);
  }
  // handle CSRF token errors here
  res.status(403);
  res.json({"message" : 'session has expired or form tampered with'});
});


// catch 404 and forward to error handler
app.use(function(req, res, next) {
  var err = new Error('Not Found');
  err.status = 404;
  next(err);
});

// error handlers
// Catch unauthorised errors
app.use(function (err, req, res, next) {
  if (err.name === 'UnauthorizedError') {
    res.status(401);
    res.json({"message" : err.name + ": " + err.message});
  }
});

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
  app.use(function(err, req, res, next) {
    res.status(err.status || 500);
    res.render('error', {
      message: err.message,
      error: err
    });
  });
}

// production error handler
// no stacktraces leaked to user
app.use(function(err, req, res, next) {
  res.status(err.status || 500);
  res.render('error', {
    message: err.message,
    error: {}
  });
});

module.exports = app;
